Add Limit Option for TaskRuns

danielhelfand opened the pull request:
Closes #163. 

This pull request adds a local flag (`--limit`) for `taskruns`. The `--limit` flag will allow a user to specify how many `taskruns` are returned from `tkn tr ls`. 

It follows the same pattern used in #185 for `pipelineruns`, but also changes the behavior for this feature for `pipelineruns`. Now, instead of returning an error when `limit` is greater than the max amount of `pipelineruns`, it will return all `pipelineruns` available.

### Usage

**Returns 5 taskruns with most recent start time**

`tkn tr ls --limit 5`

### Acceptable Input

`--limit` must be less than the total number of `taskruns` for a particular namespace and also greater than 0

### Default Case

The default case will return all `taskruns` for a particular namespace

### Local Results

![image](https://user-images.githubusercontent.com/34258252/62877976-2f1e4380-bcf6-11e9-834b-64d0d4d8b79f.png)

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

# Release Notes

```
Adds a limit flag to limit the amount of taskruns returned by the taskrun list command
```

hrishin commented:
/ok-to-test

hrishin commented on pkg/cmd/taskrun/list.go:
```diff
@@ -109,10 +127,23 @@ func list(p cli.Params, task string) (*v1alpha1.TaskRunList, error) {
 	}
 
+	// Limit after sort since taskruns are not in correct order until sorted
+	if limit > trslen {
```
Instead of returning an error, would it be nice to just return whatever objects found? :thinking:

danielhelfand replied:
My personal opinion is that it would be confusing to return something with invalid input, but, if you feel strongly about it, I can set up a conditional that will display the error message and set `limit` to 0 so it will display all `taskruns` found in the namespace.

danielhelfand replied:
If you would like me to update it, it would make sense to also do this for `pipelineruns` as well.

hrishin replied:
Well, I also felt to return an error initially and can relate the thought. However, if we step back for a moment, the user's intention is to limit the number of object to Y. So having a number of objects <=  Y anyways serves the users purpose? isn't it?  I'm missing something on this? 
 :thinking:

However, if you are not convinced then no worries at all :smile:  (Hakuna Matata)

danielhelfand replied:
Haha, I was thinking about this and came to the same conclusion. It makes sense to just return them all instead of sending an error message. 

Let me touch this up for `taskruns` and `pipelineruns`, and I'll have you give it another look.

danielhelfand replied:
Pushed up a change that will return all values when greater than max amount in a namespace and updated tests.

hrishin replied:
Thank you @danielhelfand :smile:  :+1:

hrishin commented on pkg/cmd/taskrun/list_test.go:
```diff
@@ -143,6 +144,42 @@ func TestListTaskRuns(t *testing.T) {
+		{
+			name:    "limit taskruns returned to 1",
+			command: command(t, trs, now),
+			args:    []string{"list", "-n", "foo", "-l", fmt.Sprintf("%d", 1)},
```
How about replacing `fmt.Sprintf("%d", 1)` -> `"1"`?

danielhelfand replied:
I tried using `"1"` previously in #185 and it did not work. The argument needs to be an int, so this is the way I found to pass it as a String and have it eventually interpreted as an int.

hrishin replied:
:+1:

danielhelfand commented:
@hrishin Just let me know if you would like me to squash commits.

pradeepitm12 commented:
/LGTM

danielhelfand commented:
@vdemeester @hrishin Thanks for reviewing. Fixed the CI issue and squashed commits. Just waiting on LGTM label. Let me know if there's anything else you need from me.

hrishin commented:
/lgtm
